test/library/standard/AutoMath/min-max-mix.good

```diff
@@ -0,0 +1,46 @@
+min(0:int(64), 1:int(64)) = 0 : int(64)
+min(0:uint(64), 1:uint(64)) = 0 : uint(64)
+max(0:int(64), 1:int(64)) = 1 : int(64)
+max(0:uint(64), 1:uint(64)) = 1 : uint(64)
+
+min(-9223372036854775808:int(64), 0:uint(64)) = -9223372036854775808 : int(64)
+min(-9223372036854775808:int(64), 0:uint(64)) = -9223372036854775808 : int(64)
+min(-9223372036854775808:int(64), 1:uint(64)) = -9223372036854775808 : int(64)
+min(-9223372036854775808:int(64), 18446744073709551615:uint(64)) = -9223372036854775808 : int(64)
+
+min(0:int(64), 0:uint(64)) = 0 : int(64)
+min(0:int(64), 0:uint(64)) = 0 : int(64)
+min(0:int(64), 1:uint(64)) = 0 : int(64)
+min(0:int(64), 18446744073709551615:uint(64)) = 0 : int(64)
```

Comment:
Since these are non-param, what makes them prefer the int/int overloads rather than the int/uint overloads?

Reply:
I'm a bit confused by the question. The PR adds mixed int/uint min and max to get these cases working in a nice way. What made you say they are preferring the int/int overloads?

Reply:
I see (and I made my question confusing due to writing `int/uint` where I meant `uint/uint`).  Rephrasing the question I was trying to ask:  I was expecting a mixed int/uint pair to call a uint/uint max or min since int can coerce to uint but not vice-versa.  Only now am I noticing that you've set it up so that min() returns int (since the min of any int and uint must be representable by int) and max() returns a uint(), symmetrically.

I agree that this behavior works in a pleasing way, and am trying to think about whether it'd be too confusing to someone (like me apparently) who would guess that they supported the same type mixes and therefore coercions as the math operators and one another.  Have you seen a precedent for it in other languages?  (I'll note that I am a big fan of our mixed int/uint comparison operators, but those are arguably simpler since they always return bools).

At first blush, I think I prefer the current behavior over the alternative because I'd rather be surprised by this than having a negative turn into a positive or the like.  But I think it's worth advertising pretty clearly in the OP and any characterization of this work.

Reply:
> Have you seen a precedent for it in other languages?

I don't know of any, off-hand. With a quick check, C++ gives a compilation error in that case.

<details>

``` c++
#include <cinttypes>
#include <cstdio>
#include <algorithm>

const char* tname(int8_t x)  { return "int8_t"; }
const char* tname(int16_t x) { return "int16_t"; }
const char* tname(int32_t x) { return "int32_t"; }
const char* tname(int64_t x) { return "int64_t"; }

const char* tname(uint8_t x)  { return "uint8_t"; }
const char* tname(uint16_t x) { return "uint16_t"; }
const char* tname(uint32_t x) { return "uint32_t"; }
const char* tname(uint64_t x) { return "uint64_t"; }

const char* tname(float x)  { return "float"; }
const char* tname(double x) { return "double"; }



int main() {
  int myInt = 0;
  unsigned int myUint = 1;
  printf("min(myInt, myUint) is %s\n", tname(std::min(myInt, myUint)));
  printf("max(myInt, myUint) is %s\n", tname(std::max(myInt, myUint)));

  long myLong = 0;
  unsigned long myULong = 1;
  printf("min(myLong, myULong) is %s\n", tname(std::min(myLong,myULong)));
  printf("max(myLong, myULong) is %s\n", tname(std::max(myLong,myULong)));
}
```

```
xx.cpp: In function ‘int main()’:
xx.cpp:23:54: error: no matching function for call to ‘min(int&, unsigned int&)’
   23 |   printf("min(myInt, myUint) is %s\n", tname(std::min(myInt, myUint)));
      |                                              ~~~~~~~~^~~~~~~~~~~~~~~
In file included from /usr/include/c++/11/algorithm:61,
                 from xx.cpp:3:
/usr/include/c++/11/bits/stl_algobase.h:230:5: note: candidate: ‘template<class _Tp> constexpr const _Tp& std::min(const _Tp&, const _Tp&)’
  230 |     min(const _Tp& __a, const _Tp& __b)
      |     ^~~
...
```

</details>

> At first blush, I think I prefer the current behavior over the alternative because I'd rather be surprised by this than having a negative turn into a positive or the like. 

By "current behavior", did you mean "what this PR does now" (I hope?)? I'll update the PR message.

Reply:
Two more notes about this. 

First I wanted to say that I view the min/max change here as an "obvious" generalization of the int/uint comparison operators.

Second, on `main`, `max` and `min` of mixed int/uint return a `real` which seems obviously not what we want to me.

``` chapel
var myInt: int = 1;
var myUint: uint = 2;

var z = max(myInt, myUint);
writeln("max(myInt, myUint) produces ", z.type:string);
```

main: `max(myInt, myUint) produces real(64)`

PR: `max(myInt, myUint) produces uint(64)`

Anyway for now I will assume we are in agreement but I'll be working on improving the PR message about this.

Reply:
> By "current behavior", did you mean "what this PR does now" (I hope?)?

Yeah, sorry that that was vague.

> Second, on main, max and min of mixed int/uint return a real which seems obviously not what we want to me.

Agreed.  I think if we were to do anything different than the PR's behavior, it'd be to only support min and max within a type and to have mixed int/uint cases go to the uint/uint overload, as with + and other ops.  But I'm not advocating for that.